src/Sentry/TransactionTracer.cs

```diff
@@ -0,0 +1,271 @@
+
+        private readonly ConcurrentDictionary<string, string> _tags = new();
+
+        /// <inheritdoc />
+        public IReadOnlyDictionary<string, string> Tags => _tags;
+
+        private readonly ConcurrentBag<SpanTracer> _spans = new();
+
+        /// <inheritdoc />
+        public IReadOnlyCollection<ISpanTracer> Spans => _spans;
+
+        /// <inheritdoc />
+        public bool IsFinished => EndTimestamp is not null;
+
+        /// <summary>
+        /// Initializes an instance of <see cref="Transaction"/>.
+        /// </summary>
+        public TransactionTracer(IHub hub, string name, string operation)
+        {
+            _hub = hub;
+            Name = name;
+            SpanId = SpanId.Create();
+            TraceId = SentryId.Create();
+            Operation = operation;
+        }
+
+        /// <summary>
+        /// Initializes an instance of <see cref="Transaction"/>.
+        /// </summary>
+        public TransactionTracer(IHub hub, ITransactionContext context)
+            : this(hub, context.Name, context.Operation)
+        {
+            SpanId = context.SpanId;
+            ParentSpanId = context.ParentSpanId;
+            TraceId = context.TraceId;
+            Description = context.Description;
+            Status = context.Status;
+            IsSampled = context.IsSampled;
+        }
+
+        /// <inheritdoc />
+        public void AddBreadcrumb(Breadcrumb breadcrumb) =>
+            _breadcrumbs.Add(breadcrumb);
+
+        /// <inheritdoc />
+        public void SetExtra(string key, object? value) =>
+            _extra[key] = value;
+
+        /// <inheritdoc />
+        public void SetTag(string key, string value) =>
+            _tags[key] = value;
+
+        /// <inheritdoc />
+        public void UnsetTag(string key) =>
+            _tags.TryRemove(key, out _);
+
+        internal ISpanTracer StartChild(SpanId parentSpanId, string operation)
+        {
+            var span = new SpanTracer(_hub, this, parentSpanId, operation)
+            {
+                IsSampled = IsSampled
+            };
+
+            _spans.Add(span);
+
+            return span;
+        }
+
+        /// <inheritdoc />
+        public ISpanTracer StartChild(string operation) =>
+            StartChild(SpanId, operation);
+
+        /// <inheritdoc />
+        public void Finish(SpanStatus status = SpanStatus.Ok)
+        {
+            try
+            {
+                EndTimestamp = DateTimeOffset.UtcNow;
+                Status = status;
+
+                // Client decides whether to discard this transaction based on sampling
+                _hub.CaptureTransaction(new Transaction(this));
+            }
+            finally
+            {
+                // Clear the transaction from the scope
+                _hub.ConfigureScope(scope =>
+                {
+                    if (scope.Transaction == this)
+                    {
+                        scope.Transaction = null;
+                    }
```

Comment:
should we make this Atomic? it is on Java

Reply:
@marandaneto 
I noticed that the code inside the `clearTransaction()` method is locked, but the `if (transaction == this)` part is not:

https://github.com/getsentry/sentry-java/blob/6c371b38c494ba08d0f4fbe4823cde9f378f1b5f/sentry/src/main/java/io/sentry/SentryTracer.java#L106-L111

In C# we can do `Interlocked.CompareExchange(...)` to make an atomic compare + replace operation, but that requires a direct field access...

Reply:
@Tyrrrz `scope.withTransaction` is fully locked, have a look at the `withTransaction` impl